

---
title: "Class 11"
author: "Ashley Martinez (A17891957)"
format: pdf
---

## AlphaFold Data Base

The EBI maintians the largest database of AlphaFold structure prediction models at: https://alphafold.ebi.ac.uk/

From last class before Halloween, we saw that the PDB had 244,290 (Oct 2025)

The total number of protein sequences in UniProtKB is 199,579,901

>Key Point: This is a tiny fraction of sequence space that has structual coverage(0.12%)

```{r}
244290/199579901*100
```

AFDB is attempting to address this gap...

There are two "Quality Scores" from AlphaFold one for residues (i.e. each amino acid) this is called **pLDDT** score the second is the **PAE** score. PAE measures the confidence in the relative position of two residues (i.e. a score for every pair of residues).

## Generating your own structure predictions

Figure of 5 generated HIV-PR models

![](HIVPR_DIMER_.png)

And the top ranked model colored by chain
![](HIV_Dimer_Superpose.png)

and model 5

![](HIVPR_DIMER_model5.png)

## Custom analysis of resulting models in R

Read key result files into R. The first thing I need to know is what my results directory is called (i.e. if name is different for every AlphaFold run/job)

```{r}
results_dir<-"HIVPR_dimer_23119"

#File 
pdb_files <- list.files(path=results_dir,
                        pattern="*.pdb",
                        full.names = TRUE)
#Print our PDB File names
basename(pdb_files)
```

```{r}
library(bio3d)
m1<-read.pdb(pdb_files[1])
pdbs <- pdbaln(pdb_files, fit=TRUE, exefile="msa")
```


```{r}
plot(m1$atom$b[m1$calpha],typ="l")
```

## Predicted Alignment Error for Domains

![](plots_for_HIVPR_dimer.png)
![](Seq Coverage.png)

![](plots_HIVDimer_pt2.png)

## Residue conservation from alignment file

Find the large AlphaFold alignment file

```{r}
aln_file <- list.files(path=results_dir,
                       pattern=".a3m$",
                        full.names = TRUE)
aln_file
```

How many sequences are in this alignment

```{r}
aln <- read.fasta(aln_file[1], to.upper = TRUE)
dim(aln$ali)
```

We can score residue conservation in the alignment with the conserv() function.

```{r}
sim <- conserv(aln)
pdb <- read.pdb("1hsg")
plotb3(sim[1:99], sse=trim.pdb(pdb, chain="A"),
       ylab="Conservation Score")
```

```{r}
con <- consensus(aln, cutoff = 0.9)
con$seq
```

```{r}
m1.pdb <- read.pdb(pdb_files[1])
occ <- vec2resno(c(sim[1:99], sim[1:99]), m1.pdb$atom$resno)
write.pdb(m1.pdb, o=occ, file="m1_conserv.pdb")
```

![](HIVPR_DIMER_topranked.png)






